revamp walk/1

pkoppstein opened the pull request:
Resolves #2584; also resolves #2611

thalman committed:
Fix segmentation fault when using jq in threads

In previous code nomem_handler in jv_alloc.c is called only once
and therefore the structure is not initialized for second and
following threads.

This leads to segmentation fault in multi-threading environment.

This patch moves initialization of nomem_handler out of the
pthread_once call.

thalman committed:
Test that jq works in threads

This patch implements test that searches a key in simple
json in pthread.

wader committed:
Replace NEWS with NEWS.md with more details and examples

Changes mentioned based on picking user facing changes from:
git log --oneline -r master...jq-1.6 | grep -v Merge

wader committed:
implode: Better invalid input validation and handling

Error on non-number and nan codepoint, would asserd before
Replace negative codepoint and surrogate range with unicode replacement character, would assert before

Fixes #1160

pkoppstein committed:
revamp walk/1

Resolves #2584; see also #2611

pkoppstein committed:
revamp walk/1

Resolves #2584; see also #2611

itchyny commented:
The output of `{x:0} | walk(.,1)` changes by this PR. Is it intended?

pkoppstein commented:
It is not the primary intention, which was to make the definition in terms of `map` and `map_values` as simple as possible while ensuring correct behavior when the arg was "simple" (no top-level commas). The current behavior was unintended.

Perhaps the manual should mention this example?

itchyny commented:
I just wanted to confirm that you're aware of the behavior change. Even if not mentioned in the manual, we need reasonable explanations to any behavior changes. Typical reason is that the original behavior was unintentional bug, but making the implementation simpler is also a good reason. This time, I think changing this behavior would be Okay. By the way, [in gojq I use `last`](https://github.com/itchyny/gojq/blob/v0.12.13/builtin.jq#L50-L57) to make it work like jq but I will drop it...

nicowilliams commented:
Hmm, I think we want the original behavior, yeah.

nicowilliams commented:
Or maybe I should read @2584 before commenting...

pkoppstein commented:
@nicowilliams - Perhaps it would help to review the situation w.r.t. `walk/1`.

1) The old def of `walk(f)` was written without regard to the
   possibility that f could be a sequence of expressions.  I have not
   double-checked but hopefully the official documentation
   never came remotely close to specifying what the semantics should be
   in such cases.

2)  In the case in question, I think it's generally agreed that one would want

   `{x:0} | walk(.,1)` to be equivalent to `{x:0} | walk(.), walk(1)`

    In this and other cases, jaq and the new walk are in agreement,
    and as you have no doubt noticed,  @itchyny indicated that he had to go out of his way
    to align gojq's `walk` behavior with jq's.

3) The new def elegantly resolves several important documented issues; this consideration,
   in conjunction with the fact that the new def of walk/1 is as close to a "declarative" specification of
   the desired functionality as one could achieve or want in jq, seem to me very compelling.

itchyny committed:
Rename --nul-output to --raw-output0, abort on string containing NUL

The option naming --nul-output was confusing, especially when we have a
similar option for input stream in the future (--nul-input vs --null-input).
Based on the observation of other command line tools, we rename the option
to --raw-output0. We also drop the short option -0 to avoid confusion on
introducing the NUL-delimited input option.

Unlike the other command line tools outputting file names with NUL delimiter,
jq deals with JSON, and its strings may contain NUL character. To protect
users from the risk of injection attacks, we abort the program and print an
error message before outputting strings including NUL character. Closes #2683.

nicowilliams committed:
Add JQ_FALLTHROUGH and use it to quiet warnings

nicowilliams committed:
Use -Wno-cast-function-type to quiet many warnings

nicowilliams committed:
Allow .[-1] in path expressions

pkoppstein committed:
Test negative indices in path expressions

itchyny committed:
Fail quickly if not in git repository in scripts/version

wtlangford committed:
Increment nlocals when compiling STORE_GLOBAL ops (#2000)

These are stored in vars, so we need to make sure that nlocals is large
enough to account for that

itchyny committed:
Add a regression test for #2000

itchyny committed:
Clarify the Docker image location (close #1864) (#2740)

pkoppstein committed:
manual.yml: some clarifications plus fixes for markdown bloopers (#2737)

itchyny committed:
Remove AppVeyor configuration (#2741)

itchyny committed:
Enable verbose output of make check

itchyny committed:
Validate manual schema

itchyny committed:
Fix manual source code to follow the schema definition

itchyny committed:
Update AUTHOR for the next release

itchyny committed:
Improve --help to mention all the command options

itchyny committed:
Add tests for --arg, --argjson, and similar options

itchyny committed:
Fix stderr to output string with no decoration (fix #2063)

nicowilliams committed:
Fix --stream-errors w/o --stream, and doc+test (close #2748)

itchyny committed:
Add importlib_resources dependency for older Python (fix #2753) (#2754)

itchyny committed:
Improve USE_DECNUM check position

itchyny committed:
Add decnum workflow to test building without decNumber support

itchyny committed:
Enable significand on macOS

itchyny committed:
Add checksum signature of old release artifacts (close #1999)

wader committed:
Make maintainer-mode default to disabled

This makes the standard build instructions a bit easier,
just ./configure, and also requires less tools installed (bison).

Also i think few people probably want to generate the lexer and paser code.

DavidKorczynski committed:
Add compile fuzzer for OSS-Fuzz

Signed-off-by: David Korczynski <david@adalogics.com>

nicowilliams committed:
Fix try/catch catches more than it should #1859

Close #1885, #2140, #2011, #2220, #2485, #2073

Rename the FORK_OPT opcode to TRY_BEGIN, add a TRY_END opcode, and wrap
errors when raising through a TRY_END so that they will not be caught by
the matching TRY_BEGIN.

Now a `try exp catch handler` expression generates code like:

    TRY_BEGIN handler
    <exp>
    TRY_END
    JUMP past_handler
    handler: <handler>
    past_handler:
    ...

On backtrack through TRY_BEGIN it just backtracks.

If anything past the whole thing raises when <exp> produced a value,
then the TRY_END will catch the error, wrap it in another, and
backtrack.  The TRY_BEGIN will see a wrapped error and then it will
unwrap and re-raise the error.

If <exp> raises, then TRY_BEGIN will catch the error and jump to the
handler, but the TRY_BEGIN will not stack_save() in that case, so on
raise/backtrack the TRY_BEGIN will not execute again (nor will the
TRY_END).

itchyny committed:
Improve manual in various ways (inputs, sort_by, foreach sections, etc.) (#2744)

- Add error/0 and mentions null input behavior (close #2231)
- Explain value iterator suffix syntax .foo[] (close #1047)
- Mention array slicing is also zero-based (close #2094)
- Add examples of input and inputs filters (close #2216, close #2470)
- Improve sort_by about multiple values (close #2103, close #2467, close #2474)
- Improve foreach section and simplify examples (close #1148, close #2169)
- Fix recurse/1 document on how it is identical using recurse/2 (close #2036, close #2412)
- Add non-string examples of index/1, rindex/1 (close #1422)
- Simplify the example of truncate_stream/1 (close #1736)

DavidKorczynski committed:
Add fuzzer targeting load_file

Signed-off-by: David Korczynski <david@adalogics.com>

wader committed:
Fix leaks on invalid input for strftime, strflocaltime and mktime

emanuele6 committed:
Remove all #define _GNU_SOURCE in source files

This is unnecessary since we are using AC_USE_SYSTEM_EXTENSIONS in
configure.ac, which makes the compiler always run with -D_GNU_SOURCE=1
when building for GNU/Linux.

Also, the  #define _GNU_SOURCE  in src/util.c, and in
src/inject_errors.c (--enable-inject-errors), were not guarded by a
#ifndef _GNU_SOURCE  so they were just triggering redefinition warnings.

wader committed:
Free expected value on invalid actual for test

valgrind (3.19.0) locally on macOS docker container seems to detect this leak but not CI.

$ cat modulemeta.test
modulemeta | .deps |= length
"c"
{"whatever":null,"deps":6}

$ valgrind --error-exitcode=1 --leak-check=full -q ./jq --run-tests < modulemeta.test
Test #1: 'modulemeta | .deps |= length' at line number 1
*** Insufficient results for test at line number 3: modulemeta | .deps |= length
0 of 1 tests passed (0 malformed, 0 skipped)
==14129== 482 (392 direct, 90 indirect) bytes in 1 blocks are definitely lost in loss record 7 of 8
==14129==    at 0x48407B4: malloc (vg_replace_malloc.c:381)
==14129==    by 0x13BC78: jv_mem_alloc (jv_alloc.c:122)
==14129==    by 0x1371C6: jvp_object_new (jv.c:1521)
==14129==    by 0x145840: parse_token (jv_parse.c:167)
==14129==    by 0x145840: scan (jv_parse.c:684)
==14129==    by 0x145840: jv_parser_next (jv_parse.c:793)
==14129==    by 0x1463CB: jv_parse_sized (jv_parse.c:865)
==14129==    by 0x136BD9: run_jq_tests (jq_test.c:183)
==14129==    by 0x1370FA: jq_testsuite (jq_test.c:34)
==14129==    by 0x13135C: main (main.c:582)
==14129==

emanuele6 committed:
Run  autoconf -i  in the Build step of the scanbuild workflow

Fixup from 54fef09ac39d6fb003367d65709349c43078405e

pkoppstein committed:
builtin.jq: simpler and faster transpose (#2758)

pkoppstein committed:
builtin.jq: naive abs/0

manual.yml explains that the def is naive, and mentions fabs, etc.

itchyny committed:
Reorder the command line options in the manual (#2766)

itchyny committed:
Remove deprecated --argfile option

wader committed:
Align prompt in help example

Was accidentally unaligned in #2747

Now looks like this:

Example:

	$ echo '{"foo": 0}' | jq .
	{
	  "foo": 0
	}

Before looks like this:

Example:

	 $ echo '{"foo": 0}' | jq .
	{
		"foo": 0
	}

emanuele6 committed:
Make sure to free default `@text` format on error

It seems that bison doesn't call destructors for mid-rule action
components on error, since it does not know their type.

A mid-rule action was used to allocate the "text" string used as format
by string literals without a format, which would leak on error.
This patch replaces it with a new NoFormat component of type <literal>.
Now bison will call jv_free() on that string after a syntax error.

Fixes https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=60836

  $ ./jq '"'
  jq: error: syntax error, unexpected end of file, expecting QQSTRING_TEXT or QQSTRING_INTERP_START or QQSTRING_END (Unix shell quoting issues?) at <top-level>, line 1:
  "
  jq: 1 compile error

  =================================================================
  ==1495450==ERROR: LeakSanitizer: detected memory leaks

  Direct leak of 21 byte(s) in 1 object(s) allocated from:
      #0 0x7fc21aee1359 in __interceptor_malloc /usr/src/debug/gcc/gcc/libsanitizer/asan/asan_malloc_linux.cpp:69
      #1 0x557ccb6ff608 in jv_mem_alloc src/jv_alloc.c:141

  SUMMARY: AddressSanitizer: 21 byte(s) leaked in 1 allocation(s).

emanuele6 committed:
Only care about RS when parsing with --seq

RS should only only have special meaning when parsing json-seq.
Before this fix, RS was sometimes treated as a value terminator, causing
weird results.

Bug discovered by OSS fuzz.
Ref: https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=60835

pkoppstein committed:
manual.yml: remove contingent tests

Also clarify non-prescriptive nature of some tests in jq.test

itchyny committed:
Enable gamma, drem on macOS

pkoppstein commented:
@itchyny wrote:
> By the way, [in gojq I use last](https://github.com/itchyny/gojq/blob/v0.12.13/builtin.jq#L50-L57) to make it work like jq but I will drop it...

So what should I do to get the ball rolling again?  For jq 1.7, we should at the bare minimum address the inefficiency of the 1.6 version.

I still think the version in this PR (with perhaps some elaboration in manual.yml) is the way to go, but please let me know if I should add a call to `last()`. Thanks.

itchyny committed:
Commit GPG signatures on release

itchyny committed:
Fix release job to create a pull request for signatures

emanuele6 committed:
Fix memory leak for  { $foo: bar }

{ BINDING: ExpD } wasn't freeing BINDING.

Fixes https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=60865

itchyny committed:
Fix gh pr create option on release

itchyny committed:
Fix PR creation by specifying the branch on release

itchyny committed:
Fix PR creation by using fully qualified refname on release

pkoppstein committed:
NEWS.md: tweaks

correct grammar, add attributions, clarify abs

owenthereal committed:
Redirect questions to Discord

We now have an official Discord server and most maintainers are hanging
out there. It would be a good idea to redirect questions to Discord.

owenthereal committed:
Update `bug_report.md` with Discord link

owenthereal committed:
Backfill with references in NEWS.md (#2788)

Backfill with references to PRs & issues in NEWS.md

emanuele6 committed:
Fix typo in manual: "-seq" => "--seq"

itchyny committed:
Redesign website (#2628)

* Bump up Bootstrap to v5.3.1, Bootstrap Icon to v1.10.5.
* Use autoComplete.js to drop dependency on jQuery and typeahead.js.
* Support dark mode.
* New svg logo and icon with responsive color mode support.
* Normalize section ids to lower kebab-case for easiness of linking.
* Use relative paths for links for local development (--root /output).
* Various markup cleanups and accessibility improvements.

itchyny commented on src/builtin.jq:
```diff
@@ -249,17 +249,18 @@ def bsearch($target):
       end
   end;
-
+#
```
I prefer this line to be an empty line rather than an empty comment.

pkoppstein replied:
OK, but someone should probably run: `s/^# *$//`

itchyny commented:
Please rebase this PR, thanks.

pkoppstein committed:
revamp walk/1: rebase, add test cases

Note that according to the new def:

`{x:0} | walk(.,1)` is equivalent to: {x:0} | walk(.), walk(1)

pkoppstein committed:
Merge branch 'walk' of https://github.com/pkoppstein/jq into walk

pkoppstein commented:
@itchyny wrote:
> Please rebase this PR, thanks.

I did, but somehow a "merge" was added as well.  Feel free to "squash" as desired.

itchyny commented:
Could you rebase the commit onto the latest remote master branch? This PR includes the commits ahead and I can't review the diff now.

pkoppstein commented:
Given the difficulty merging this PR, I'm closing it in favor of https://github.com/jqlang/jq/pull/2795

nicowilliams commented:
@pkoppstein if you would only just use your terminal application and the git command you wouldn't have such difficulties and you wouldn't cause so many duplicate PRs.

pkoppstein commented:
@nicowilliams wrote:
> use your terminal application and the git command 

In brief, I did, but itchyny said he could not use the result of a simple "Rebase" and "Push".

In long - well, let me know if and when you would like further details.
